### PR TITLE
Replace locale in redirect_uri parameter in translation menu

### PR DESCRIPTION
**Pull request by chives**

### CODE REVIEW

File: EventListener/TranslationLocaleMenuListener.php [hunk 1]
```diff
@@ -108,4 +128,71 @@ private function populateTranslationLocaleMenu(Item $menu)
             $menu->addChild($localeItem);
         }
     }
+
+    /**
+     * @param string $redirectUri
+     * @param $locale
+     * @return array
+     */
+    private function rebuildRedirectUri($redirectUri, $locale)
```
Comment (rn0) [inline EventListener/TranslationLocaleMenuListener.php]: not used

File: EventListener/TranslationLocaleMenuListener.php [hunk 2]
```diff
@@ -108,4 +128,57 @@ private function populateTranslationLocaleMenu(Item $menu)
...
+    private function replaceUriPath($uri, $newPath)
+    {
+        $uriParts = parse_url($uri);
+
+        return sprintf('%s://%s%s%s',
+            isset($uriParts['scheme']) ? $uriParts['scheme'] : 'http',
+            isset($uriParts['host']) ? $uriParts['host'] : $this->request->server['HTTP_HOST'],
+            $newPath,
+            isset($uriParts['query']) ? ('?' . urldecode($uriParts['query'])) : ''
+        );
```
Comment (JarekW) [inline EventListener/TranslationLocaleMenuListener.php]: maybe [http_build_url](http://php.net/manual/en/function.http-build-url.php#example-4504) will do the trick

**Comment by JarekW** [comment]
why link must be absolute and url are parsed and generated again?

File: EventListener/TranslationLocaleMenuListener.php (hunk 2 above)
Comment (chives) [inline EventListener/TranslationLocaleMenuListener.php]: Unfortunately it's not a built-in function but requires pecl_http extension

Comment (JarekW) [inline EventListener/TranslationLocaleMenuListener.php]: ok

**Comment by chives** [comment]
``redirect_uri`` does not have to be absolute but it's passed this way from many places in https://github.com/fsi-open/admin-bundle, https://github.com/fsi-open/datagrid-bundle and https://github.com/fsi-open/datasource-bundle and it has to be rebuilt (in the same absolute form) when translation locale is changed in order to redirect to the new locale after successful action such as saving an edit form.

**Comment by JarekW** [comment]
ok, :+1:

File: EventListener/TranslationLocaleMenuListener.php (hunk 2 above)
Comment (bartekgruszka) [inline EventListener/TranslationLocaleMenuListener.php]: I am not 100% sure if it's possible, but I think you should consider using a [RequestContext](http://api.symfony.com/2.5/Symfony/Component/Routing/RequestContext.html) and then [UrlGenerator:;generate](http://api.symfony.com/2.6/Symfony/Component/Routing/Generator/UrlGenerator.html#method_generate)

**Comment by chives** [comment]
In order to make this change actually work we will need to change what is passed as ``redirect_uri`` parameter (i.e. app.request.requestUri instead of app.request.uri)
https://github.com/search?l=&q=redirect_uri+user%3Afsi-open&ref=advsearch&type=Code&utf8=%E2%9C%93

**Comment by bartekgruszka** [comment]
:+1:

**Comment by chives** [comment]
Related PRs - https://github.com/fsi-open/admin-bundle/pull/184, https://github.com/fsi-open/datagrid/pull/59

**Comment by chives** [comment]
@fsi-open

**Comment by chives** [comment]
@fsi-open rebased

**Comment by szymach** [comment]
:+1:

**Comment by bartekgruszka** [comment]
:+1:

**Comment by chives** [comment]
@fsi-open

File: EventListener/TranslationLocaleMenuListener.php [hunk 3]
```diff
@@ -9,6 +9,9 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\RequestStack;
 use Symfony\Component\Intl\Intl;
+use Symfony\Component\Routing\Exception\ResourceNotFoundException;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Routing\Router;
```
Comment (szymach) [inline EventListener/TranslationLocaleMenuListener.php]: nie powinno być RouterInterface?

Comment (chives) [inline EventListener/TranslationLocaleMenuListener.php]: ``RouterInterface`` hasn't required ``matchRequest()`` method used [here](https://github.com/fsi-open/admin-translatable-bundle/pull/35/files#diff-c14e6c780cf26b1b83c6da9bbd2a9fa7R161)

**Comment by szymach** [comment]
:+1:

**Comment by rn0** [comment]
:+1:

**Pull request merged**